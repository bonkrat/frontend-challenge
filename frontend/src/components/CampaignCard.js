import React from "react";
import PropTypes from "prop-types";
import FundedBar from "./FundedBar";
import FundButton from "./FundButton";
import Backers from "./Backers";
import { Link } from "react-router-dom";
import styles from "./CampaignCard.styles";

/**
 * Links children to a product
 */
export const LinkToProduct = ({ id, style, children }) => (
  <Link to={"/" + id} style={style}>
    {children}
  </Link>
);

LinkToProduct.propTypes = {
  /**
   * The ID of the product to link to.
   */
  id: PropTypes.string,
  /**
   * The style of the link.
   */
  style: PropTypes.object
};

/**
 * A card rendering a campaign.
 *
 * @param {object} props
 */
const CampaignCard = ({
  size = ["flex-1"],
  hero,
  id,
  title,
  description,
  image,
  goal,
  funds,
  creator,
  backers
}) => (
  <div className={styles.card(size)}>
    <LinkToProduct
      id={id}
      style={{
        backgroundImage: `url("http://localhost:4000/images/${image}")`,
        backgroundPosition: "center",
        backgroundSize: "cover"
      }}
    >
      <div className={styles.image(hero)} title="name"></div>
    </LinkToProduct>
    <div className={styles.cardContents(hero)}>
      <div className={styles.title}>
        <LinkToProduct id={id}>
          <div className={styles.titleText}>{title}</div>
        </LinkToProduct>
        {creator && (
          <div className={styles.creator(hero)}>Created by {creator.name}</div>
        )}
        <p className={styles.description(hero)}>{description}</p>
      </div>
      <Backers backers={backers} />
      <div className={styles.fund}>
        <FundedBar className="w-3/4" goal={goal} funds={funds} />
        <LinkToProduct id={id}>
          <FundButton className="w-1/4" />
        </LinkToProduct>
      </div>
    </div>
  </div>
);

CampaignCard.propTypes = {
  /**
   * An array of tailwind classes describing the size of the card.
   */
  size: PropTypes.array,
  /**
   * If true, the card is rendered larger as a 'hero' card.
   */
  hero: PropTypes.bool,
  /**
   * The id of the campaign.
   */
  id: PropTypes.string,
  /**
   * The title of the campaign.
   */
  title: PropTypes.string,
  /**
   * The description of the campaign.
   */
  description: PropTypes.string,
  /**
   * The URL to the image for the campaign.
   */
  image: PropTypes.string,
  /**
   * The goal amount of the campaign.
   */
  goal: PropTypes.number,
  /**
   * The amount of funding the campaign has raised.
   */
  funds: PropTypes.number,
  /**
   * The creator of the campaign.
   */
  creator: PropTypes.object,
  /**
   * Backers of the campaign.
   */
  backers: PropTypes.array
};

export default CampaignCard;
