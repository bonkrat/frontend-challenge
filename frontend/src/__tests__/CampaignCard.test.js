import React from "react";
import { shallow } from "enzyme";
import CampaignCard, { LinkToProduct } from "../components/CampaignCard";
import toJson from "enzyme-to-json";

describe("CampaignCard", () => {
  const props = {
    size: ["flex-2"],
    hero: false,
    id: "fooid",
    description: "foodescription",
    image: "fooimage",
    goal: 100,
    funds: 25,
    creator: {
      name: "fooname"
    },
    backers: [
      {
        id: "foo",
        name: "fooname",
        avatar: "fooavatar"
      },
      {
        id: "bar",
        name: "barname",
        avatar: "baravatar"
      }
    ]
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<CampaignCard {...props} />);
  });

  it("renders correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders correct styles when the card is a hero card", () => {
    wrapper.setProps({
      ...props,
      hero: true
    });
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
describe("LinkToProduct", () => {
  it("renders correctly", () => {
    const wrapper = shallow(
      <LinkToProduct id="foo" style={{ color: "green" }}>
        Foo Child
      </LinkToProduct>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
