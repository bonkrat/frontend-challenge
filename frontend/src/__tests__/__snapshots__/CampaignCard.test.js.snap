// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampaignCard renders correct styles when the card is a hero card 1`] = `
<div
  className="flex flex-grow px-2 mb-4 flex-2"
>
  <LinkToProduct
    id="fooid"
    style={
      Object {
        "backgroundImage": "url(\\"http://localhost:4000/images/fooimage\\")",
        "backgroundPosition": "center",
        "backgroundSize": "cover",
      }
    }
  >
    <div
      className="h-auto flex-none bg-cover rounded-t lg:rounded-t-none lg:rounded-l text-center overflow-hidden lg:w-64 md:w-56 w-48"
      title="name"
    />
  </LinkToProduct>
  <div
    className="border-b border-l lg-border-l-0 bg-white rounded-b lg:rounded-b-none lg: rouned-r flex flex-col justify-between leading-normal p-4 w-full md:p-8"
  >
    <div
      className="mb-8"
    >
      <LinkToProduct
        id="fooid"
      >
        <div
          className="text-gray-900 font-bold text-lx hover:text-purple-900 cursor-pointer"
        />
      </LinkToProduct>
      <div
        className="text-sm mb-4 text-gray-500 text-base"
      >
        Created by 
        fooname
      </div>
      <p
        className="text-gray-700 text-sm pb-12 md:text-base"
      >
        foodescription
      </p>
    </div>
    <Backers
      backers={
        Array [
          Object {
            "avatar": "fooavatar",
            "id": "foo",
            "name": "fooname",
          },
          Object {
            "avatar": "baravatar",
            "id": "bar",
            "name": "barname",
          },
        ]
      }
      size={6}
    />
    <div
      className="flex justify-between"
    >
      <FundedBar
        className="w-3/4"
        funds={25}
        goal={100}
      />
      <LinkToProduct
        id="fooid"
      >
        <FundButton
          className="w-1/4"
          loading={false}
          onClick={[Function]}
        />
      </LinkToProduct>
    </div>
  </div>
</div>
`;

exports[`CampaignCard renders correctly 1`] = `
<div
  className="flex flex-grow px-2 mb-4 flex-2"
>
  <LinkToProduct
    id="fooid"
    style={
      Object {
        "backgroundImage": "url(\\"http://localhost:4000/images/fooimage\\")",
        "backgroundPosition": "center",
        "backgroundSize": "cover",
      }
    }
  >
    <div
      className="h-auto flex-none bg-cover rounded-t lg:rounded-t-none lg:rounded-l text-center overflow-hidden w-48"
      title="name"
    />
  </LinkToProduct>
  <div
    className="border-b border-l lg-border-l-0 bg-white rounded-b lg:rounded-b-none lg: rouned-r flex flex-col justify-between leading-normal p-4 w-full border-t"
  >
    <div
      className="mb-8"
    >
      <LinkToProduct
        id="fooid"
      >
        <div
          className="text-gray-900 font-bold text-lx hover:text-purple-900 cursor-pointer"
        />
      </LinkToProduct>
      <div
        className="text-sm mb-4 text-gray-500"
      >
        Created by 
        fooname
      </div>
      <p
        className="text-gray-700 text-sm"
      >
        foodescription
      </p>
    </div>
    <Backers
      backers={
        Array [
          Object {
            "avatar": "fooavatar",
            "id": "foo",
            "name": "fooname",
          },
          Object {
            "avatar": "baravatar",
            "id": "bar",
            "name": "barname",
          },
        ]
      }
      size={6}
    />
    <div
      className="flex justify-between"
    >
      <FundedBar
        className="w-3/4"
        funds={25}
        goal={100}
      />
      <LinkToProduct
        id="fooid"
      >
        <FundButton
          className="w-1/4"
          loading={false}
          onClick={[Function]}
        />
      </LinkToProduct>
    </div>
  </div>
</div>
`;

exports[`LinkToProduct renders correctly 1`] = `
<Link
  style={
    Object {
      "color": "green",
    }
  }
  to="/foo"
>
  Foo Child
</Link>
`;
